src/read.rs

```diff
@@ -686,10 +691,28 @@ impl<R: Read + Seek> ZipArchive<R> {
                 if let Some(p) = outpath.parent() {
                     Self::make_writable_dir_all(p)?;
                 }
+                /* FIXME: what bug does this code fix?
+                 * https://github.com/zip-rs/zip2/commit/8715d936cbd3e0fdc2e778b70d8a02e2dfc16fc2
+                 * doesn't specify! */
                 if file.is_symlink() && (cfg!(unix) || cfg!(windows)) {
                     let mut target = Vec::with_capacity(file.size() as usize);
                     file.read_exact(&mut target)?;
-                    let target_path: PathBuf = directory.as_ref().join(OsString::try_from(target)?);
+                    #[cfg(unix)]
+                    let target_str = OsString::from_vec(target);
+                    #[cfg(windows)]
+                    let target_str: OsString = {
+                        let chunks = target.chunks_exact(2);
+                        assert!(
+                            chunks.remainder().is_empty(),
+                            "windows utf-16 strings should be divisible by 2"
+                        );
+                        let target_wide: Vec<u16> = chunks
+                            .into_iter()
+                            .map(|c| u16::from_le_bytes(c.try_into().unwrap()))
```

Comment:
This will panic if a character code point is above 0xFFFF, because that has to be encoded in UTF-16 as a surrogate pair.